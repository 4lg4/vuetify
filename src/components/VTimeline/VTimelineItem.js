import { inject as RegistrableInject } from '../../mixins/registrable'

import Colorable from '../../mixins/colorable'

import VIcon from '../VIcon'

export default {
  name: 'v-timeline-item',

  mixins: [Colorable, RegistrableInject('timeline', 'v-timeline', 'v-timeline-item')],

  props: {
    icon: {
      type: String,
      default: 'event'
    },
    noIcon: {
      type: Boolean,
      default: false
    },
    lineColor: { type: String },
    iconFillColor: { type: String }
  },

  computed: {
    classes () {
      if (this.lineColor) {
        return this.addTextColorClassChecks({}, this.lineColor)
      }

      return ''
    },
    iconFillColorClasses () {
      return this.iconFillColor ? this.iconFillColor : this.$parent.$props.iconFillColor ? this.$parent.$props.iconFillColor : null
    }
  },

  methods: {
    genBody () {
      return this.$createElement(
        'div',
        {
          ref: 'body',
          class: 'timeline__item--body'
        },
        this.$slots.item
      )
    },
    genIconHeader () {
      const icon = this.noIcon || this.$parent.$props.noIcon ? null : this.$slots.icon || this.$createElement(VIcon, this.icon)
      return this.$createElement(
        'div',
        {
          staticClass: 'timeline__item--head'
        },
        [
          this.$createElement(
            'div',
            {
              staticClass: 'timeline__item--head-icon',
              class: this.iconFillColorClasses
            },
            [icon]
          )
        ]
      )
    }
  },

  render (h) {
    const children = []

    children.push(this.genIconHeader())
    children.push(this.genBody())

    return h(
      'li',
      {
        staticClass: 'timeline__item',
        class: this.classes
      },
      children
    )
  }
}
