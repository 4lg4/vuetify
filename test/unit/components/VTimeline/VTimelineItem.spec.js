import { test } from '@/test'
import VTimelineItem from '@/components/VTimeline/VTimelineItem'

test('VTimelineItem.js', ({ mount }) => {
  function timelineItemProvideData () {
    const provideData = {
      provide: {
        iconParent: 'event',
        iconSizeParent: 24,
        noIconParent: false,
        hoverParent: false,
        raisedParent: false,
        lineColorParent: 'grey lighten-2',
        circleFillColorParent: 'white',
        circleOutlineColorParent: 'grey lighten-2',
        lineSizeParent: 8,
        hideCircleOutlineParent: false
      }
    }
    return provideData.provide
  }
  it('renders a li with class timeline__item', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      provide
    })

    const li = wrapper.first('li')
    expect(li.is('li')).toBe(true)
    expect(li.hasClass('timeline__item')).toBe(true)
  })

  it('should render timeline body with timeline--raised class when raised prop', () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        raised: true
      },
      provide
    })

    const div = wrapper.find('.timeline__item--body')[0]
    expect(div.hasClass('timeline--raised')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render timeline body with timeline--hover class when hover prop', () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        hover: true
      },
      provide
    })
    const div = wrapper.find('.timeline__item--body')[0]
    expect(div.hasClass('timeline--hover')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a custom icon', async () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        icon: 'list'
      },
      provide
    })
    const div = wrapper.find('.timeline__icon')[0]
    expect(div.text()).toBe('list')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a custom icon without interference from parent', () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        iconParent: 'event',
        icon: 'list'
      },
      provide
    })
    const div = wrapper.find('.timeline__icon')[0]
    expect(div.text()).toBe('list')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a custom icon size', () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        iconSize: 48
      },
      provide
    })
    const div = wrapper.find('.timeline__icon .icon')[0]
    expect(div.hasStyle('fint-size', '48px')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a custom size without interference from parent', () => {
    const provide = timelineItemProvideData()

    const wrapper = mount(VTimelineItem, {
      propsData: {
        iconSizeParent: 48,
        iconSize: 58
      },
      provide
    })
    const div = wrapper.find('.timeline__icon .icon')[0]
    expect(div.hasStyle('fint-size', '58px')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a no icon', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        noIcon: true
      },
      provide
    })

    const div = wrapper.find('.timeline__icon .icon')
    expect(div.length).toBe(0)

    wrapper.setProps({ noIcon: false })
    const div2 = wrapper.find('.timeline__icon .icon')
    expect(div2.length).toBe(1)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a no icon without parent interference', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        noIconParent: false,
        noIcon: true
      },
      provide
    })

    const div = wrapper.find('.timeline__icon .icon')
    expect(div.length).toBe(0)

    wrapper.setProps({ noIconParent: true, noIcon: false })
    const div2 = wrapper.find('.timeline__icon .icon')
    expect(div2.length).toBe(1)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a cutom line color', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        lineColor: 'red lighten-1'
      },
      provide
    })

    const div = wrapper.find('.timeline__line')[0]

    expect(div.hasClass('red--text', 'text--lighten-1')).toBe(true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a cutom line color without parent interference', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        lineColorParent: 'red lighten-2',
        lineColor: 'red lighten-1'
      },
      provide
    })

    const div = wrapper.find('.timeline__line')[0]

    expect(div.hasClass('red--text', 'text--lighten-1')).toBe(true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have cutom circle outline color', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        circleOutlineColor: 'red'
      },
      provide
    })
    const div = wrapper.find('.timeline__item--head')[0]
    expect(div.hasClass('red--text')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have cutom circle outline color without parent interference', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        circleOutlineColorParent: 'blue',
        circleOutlineColor: 'red'
      },
      provide
    })
    const div = wrapper.find('.timeline__item--head')[0]
    expect(div.hasClass('red--text')).toBe(true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have cutom circle fill color', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        circleFillColor: 'red'
      },
      provide
    })
    const div = wrapper.find('.timeline__item--head')[0]
    expect(div.hasClass('red')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have cutom circle fill color without parent interference', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        circleFillColorParent: 'blue',
        circleFillColor: 'red'
      },
      provide
    })
    const div = wrapper.find('.timeline__item--head')[0]
    expect(div.hasClass('red')).toBe(true)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should hide circle outline', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        hideCircleOutline: true
      },
      provide
    })
    const div = wrapper.find('.timeline__icon')[0]
    expect(div.hasStyle('box-shadow', 'none')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should hide circle outline without parent interference', () => {
    const provide = timelineItemProvideData()
    const wrapper = mount(VTimelineItem, {
      propsData: {
        hideCircleOutlineParent: false,
        hideCircleOutline: true
      },
      provide
    })
    const div = wrapper.find('.timeline__icon')[0]
    expect(div.hasStyle('box-shadow', 'none')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
