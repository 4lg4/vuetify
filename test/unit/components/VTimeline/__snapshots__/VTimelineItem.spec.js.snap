// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTimelineItem.js should have a custom icon 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        list
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a custom icon size 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 48px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a custom icon without interference from parent 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        list
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a custom size without interference from parent 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 58px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a cutom line color 1`] = `

<li class="timeline__item">
  <div class="timeline__line red--text text--lighten-1"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a cutom line color without parent interference 1`] = `

<li class="timeline__item">
  <div class="timeline__line red--text text--lighten-1"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a no icon 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have a no icon without parent interference 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have cutom circle fill color 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head red grey--text text--lighten-2">
    <div class="timeline__icon red grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have cutom circle fill color without parent interference 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head red grey--text text--lighten-2">
    <div class="timeline__icon red grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have cutom circle outline color 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white red--text">
    <div class="timeline__icon white red--text"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should have cutom circle outline color without parent interference 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white red--text">
    <div class="timeline__icon white red--text"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should hide circle outline 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: none;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should hide circle outline without parent interference 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: none;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should render timeline body with timeline--hover class when hover prop 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body timeline--hover">
    </div>
  </div>
</li>

`;

exports[`VTimelineItem.js should render timeline body with timeline--raised class when raised prop 1`] = `

<li class="timeline__item">
  <div class="timeline__line grey--text text--lighten-2"
       style="width: 8px; margin-left: -0.66em;"
  >
  </div>
  <div class="timeline__item--head white grey--text text--lighten-2">
    <div class="timeline__icon white grey--text text--lighten-2"
         style="box-shadow: currentColor 0px 0px 0px 8px;"
    >
      <i aria-hidden="true"
         class="icon material-icons"
         style="font-size: 24px;"
      >
        event
      </i>
    </div>
  </div>
  <div class="timeline__item--body__container">
    <div class="timeline__item--body timeline--raised">
    </div>
  </div>
</li>

`;
